module/zfs/dmu_object.c

```diff
@@ -61,6 +61,7 @@ dmu_object_alloc_dnsize(objset_t *os, dmu_object_type_t ot, int blocksize,
 	boolean_t restarted = B_FALSE;
 	uint64_t *cpuobj = NULL;
 	int dnodes_per_chunk = 1 << dmu_object_alloc_chunk_shift;
+	int error;
 
 	kpreempt_disable();
 	cpuobj = &os->os_obj_next_percpu[CPU_SEQID %
```

Comment:
The multi-threaded allocator, which introduced the atomics, is only responsible for making a best effort to so concurrent allocations _usually_ occur in different dbufs.  This minimizes contention of the dbuf lock.  As you noted since preemption isn't disabled over the entire allocation it is possible two CPUs could try and allocate the same object.  Or in the case of this PR different, but overlapping objects.  The `dnode_hold_impl()` function is intended to authoritatively resolve these races safely when they occur.